flink-runtime/src/main/java/org/apache/flink/runtime/util/ScheduledFutures.java

```diff
@@ -0,0 +1,58 @@
+package org.apache.flink.runtime.util;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadFactory;
+import org.apache.flink.shaded.guava18.com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+/**
+ * Utils related to {@link ScheduledFuture scheduled futures}.
+ */
+public class ScheduledFutures {
+
+	private static final ThreadFactory THREAD_FACTORY = new ThreadFactoryBuilder()
+		.setNameFormat("Flink-Scheduled-Future-SafeGuard")
+		.setDaemon(true)
+		.build();
+
+	/**
+	 * Guard {@link ScheduledFuture} with scheduled future uncaughtException handler, because
+	 * {@link java.util.concurrent.ScheduledExecutorService} does not respect the one assigned to
+	 * executing {@link Thread} instance.
+	 *
+	 * @param scheduledFuture Scheduled future to guard.
+	 * @param uncaughtExceptionHandler Handler to call in case of uncaught exception.
+	 * @param <T> Type the future returns.
+	 * @return Future with handler.
+	 */
+	public static <T> ScheduledFuture<T> withUncaughtExceptionHandler(
+		ScheduledFuture<T> scheduledFuture,
+		Thread.UncaughtExceptionHandler uncaughtExceptionHandler) {
+		final Thread safeguardThread = THREAD_FACTORY.newThread(() -> {
```

Comment:
Why do we create a new thread for this, instead of using an executor which already has a thread pool? We can still wrap the Runnable if we want.

Reply:
We need a separate thread to `get()` the result of `ScheduledFuture`, which is blocking op. 

Wrapping the scheduled `Runnable` (in the one that try/catches throwable and calls UEH) instead of `ScheduledFuture` should get us the same result.

I'm OK with either approach. Which one do you prefer?

Reply:
I still prefer using the existing thread pools, i.e. wrapping the Runnable.

Reply:
Reimplemented using Runnable wrapper.